File: frontend/lib/invoice/tax-breakdown.ts
import { AdminOrderDetail } from '@/lib/types'
import { itemSubtotal } from '@/lib/admin/order-documents'

export interface TaxBreakdownRow {
  ratePercent: number
  amountInclusive: number
  consumptionTax: number
}

export function taxFromInclusive(amountInclusive: number, ratePercent: number): {
  consumptionTax: number
  amountExcludingTax: number
} {
  const amount = Math.max(0, Math.round(amountInclusive))
  const consumptionTax = Math.round((amount * ratePercent) / (100 + ratePercent))
  return { consumptionTax, amountExcludingTax: amount - consumptionTax }
}

export function buildOrderTaxBreakdown(
  order: AdminOrderDetail,
  defaultTaxRate: number
): TaxBreakdownRow[] {
  const subtotal = itemSubtotal(order)
  const shipping = order.shipping_fee || 0
  const packaging = order.packaging_fee || 0
  const fee = order.payment_fee || 0
  const discount = order.discount_amount || 0

  const grouped = new Map<number, number>()
  const add = (amount: number) => {
    if (amount === 0) return
    grouped.set(defaultTaxRate, (grouped.get(defaultTaxRate) || 0) + amount)
  }

  add(subtotal)
  add(shipping)
  add(packaging)
  add(fee)
  if (discount > 0) add(-discount)

  return Array.from(grouped.entries())
    .sort(([a], [b]) => b - a)
    .map(([ratePercent, amountInclusive]) => {
      const { consumptionTax } = taxFromInclusive(amountInclusive, ratePercent)
      return { ratePercent, amountInclusive, consumptionTax }
    })
    .filter((row) => row.amountInclusive !== 0)
}

File: backend/services/tax_breakdown.py
"""Tax breakdown from tax-included (税込) amounts."""

from __future__ import annotations

from dataclasses import dataclass

import schemas


@dataclass
class TaxSlice:
    rate_percent: int
    amount_inclusive: int


def tax_from_inclusive(amount_inclusive: int, rate_percent: int) -> tuple[int, int]:
    """Return (consumption_tax, amount_excluding_tax) from tax-included amount."""
    amount = max(0, int(round(amount_inclusive)))
    tax = int(round(amount * rate_percent / (100 + rate_percent)))
    return tax, amount - tax


def build_tax_breakdown_rows(slices: list[TaxSlice]) -> list[schemas.TaxBreakdownRowOut]:
    grouped: dict[int, int] = {}
    for s in slices:
        if s.amount_inclusive <= 0:
            continue
        grouped[s.rate_percent] = grouped.get(s.rate_percent, 0) + s.amount_inclusive

    rows: list[schemas.TaxBreakdownRowOut] = []
    for rate in sorted(grouped.keys(), reverse=True):
        inclusive = grouped[rate]
        tax, _ex = tax_from_inclusive(inclusive, rate)
        rows.append(
            schemas.TaxBreakdownRowOut(
                rate_percent=rate,
                amount_inclusive=inclusive,
                consumption_tax=tax,
            )
        )
    return rows


def order_tax_slices(
    *,
    items_subtotal: int,
    shipping_fee: int,
    packaging_fee: int,
    payment_fee: int,
    discount_amount: int,
    default_tax_rate: int,
) -> list[TaxSlice]:
    """Build tax slices for an order (all at default rate for now)."""
    rate = default_tax_rate
    slices: list[TaxSlice] = []
    for amount in (items_subtotal, shipping_fee, packaging_fee, payment_fee):
        if amount > 0:
            slices.append(TaxSlice(rate_percent=rate, amount_inclusive=amount))
    if discount_amount > 0:
        slices.append(TaxSlice(rate_percent=rate, amount_inclusive=-discount_amount))
    return slices
